bgzip.c

```diff
@@ -103,6 +106,83 @@ static int confirm_filename(int *is_forced, const char *name, const char *ext)
     return ask_yn();
 }
 
+/* getfilespec - get file status data
+   path        - file path for which status to be retrieved
+   filestatus  - pointer to status structure in which the data to be stored
+   returns 0 on success and -1 on failure
+*/
+static int getfilespec(const char *path, void *filestatus)
+{
+#ifdef _WIN32
+    struct _stat *status = (struct _stat*)filestatus;
+#else
+    struct stat *status = (struct stat*)filestatus;
+#endif //_WIN32
+
+    if (!path || !status) { //invalid
+        return -1;
+    }
+    if (!strcmp(path, "-")) {   //cant get / set for stdin/out, return success
+        return 0;
+    }
+#ifdef _WIN32
+    if (_stat(path, status) < 0) {
+        return -1;
+    }
+#else
+    if (stat(path, status) < 0) {
+        return -1;
+    }
+#endif //_WIN32
+    return 0;
+}
+
+/* setfilespec - set file status data
+   path        - file path for which status to be set
+   filestatus  - pointer to status structure in which the data is present
+   returns 0 on success and -1 on failure
+   sets only the modification time as of now.
+*/
+static int setfilespec(const char *path, const void *filestatus)
+{
+#ifdef _WIN32
+    struct _stat *status = (struct _stat*)filestatus;
+    struct _utimbuf tval;
+#else
+    struct stat *status = (struct stat*)filestatus;
+    struct timeval tval[2];
+#endif //_WIN32
+
+    if (!path || !filestatus) { //invalid
+        return -1;
+    }
+    if (!strcmp(path, "-")) {   //cant get / set for stdin/out, return success
+        return 0;
+    }
+
+#ifdef _WIN32
+    //time upto sec - access & modification time
+    tval.actime = status->st_atime;
+    tval.modtime = status->st_mtime;
+    if (_utime(path, &tval) < 0) {
+        fprintf(stderr, "[bgzip] Failed to set file specifications.\n");
+        return -1;
+    }
+#else
+    //time upto usec - access time
+    tval[0].tv_sec = status->st_atime;
+    tval[0].tv_usec = status->st_atimensec / 1000;
+    //time upto usec - modification time
+    tval[1].tv_sec = status->st_mtime;
+    tval[1].tv_usec = status->st_mtimensec / 1000;
+    if (utimes(path, &tval[0]) < 0) {
+        fprintf(stderr, "[bgzip] Failed to set file specifications.\n");
+        return -1;
```

Comment:
Do we need nanosec accuracy?

It's thwart with problems.  See the myriad of code in https://github.com/kunpengcompute/gzip/blob/6a3ebab7e475fb5ca9b01c344eb45c27ea81ef89/lib/stat-time.h#L66

It also doesn't build on linux even unless we define `_XOPEN_SOURCE` or similar.  For example:

```
$ cat _.c
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char **argv) {
  struct stat sb;
  if (stat(argv[1], &sb) == 0)
    printf("%ld %ld\n", sb.st_mtime, sb.st_mtimensec);

  return 0;
}

$ cc _.c
_.c: In function 'main':
_.c:11:41: error: 'struct stat' has no member named 'st_mtimensec'; did you mean 'st_mtim'?
     printf("%ld %ld\n", sb.st_mtime, sb.st_mtimensec);
                                         ^~~~~~~~~~~~
                                         st_mtim

```

While we could do nanosecond precise, frankly it's probably sufficient to just squash that to zero and just do second precision.  That will make things far more portable without the need for lots of ifdefs.

Reply:
Also, I note that `cc -D_XOPEN_SOURCE=600` works, but `cc -D_XOPEN_SOURCE=700` does not.  We get away with it in htslib because we add 600 to the config.h file.  However I'm not sure we want to rely on something which is already known to have been removed in newer XOpen specifications.

Reply:
Thinking more, if we did want to keep nanoseconds, then we could do a `#if defined(_XOPEN_SOURCE) && _XOPEN_SOURCE == 600` ... or similar.  However I'm still not convnced of the necessity of it being this fine grained.

Reply:
changed to use the same stat structure for all OS.
changed to set time value with second precision on all cases.